File: 22/cpp/main.cpp
#include <iostream>
#include <numeric>
#include <stack>
#include <string>
#include <vector>

class Solution {
public:
    // Stack and result vector as member variables
    std::stack<std::string> stack;
    std::vector<std::string> result;

    std::vector<std::string> generateParenthesis(int n) {
        backtrack(n, 0, 0);  // Call the backtrack function
        return result;
    }

private:
    void backtrack(int n, int open, int closed) {
        if (open == n && closed == n) {
            // Pop elements from the stack and concatenate
            std::string to_append = "";
            std::stack<std::string> temp_stack = stack;  // Temporary copy of the stack
            while (!temp_stack.empty()) {
                to_append = temp_stack.top() + to_append;
                temp_stack.pop();
            }
            result.emplace_back(to_append);
            return;
        }

        if (open < n) {
            stack.push("(");
            backtrack(n, open + 1, closed);
            stack.pop();  // Backtrack, remove the last added element
        }

        if (closed < open) {
            stack.push(")");
            backtrack(n, open, closed + 1);
            stack.pop();  // Backtrack, remove the last added element
        }
    }
};

int main() {
    Solution sol;

    // Test case X: Y
    RETURN_TYPE resultX = sol.FUNCTION_NAME(Y);
    std::cout << "Test X: " << resultX << std::endl;

    return 0;
}

File: 22/python3/main.py
#!/usr/bin/env python3

class Solution:
    def __init__(self):
        # Initialize the stack and result as member variables
        self.stack = []
        self.result = []

    def generateParenthesis(self, n: int):
        self.backtrack(n, 0, 0)  # Start backtracking with open and closed counts at 0
        return self.result

    def backtrack(self, n, open_count, closed_count):
        # If we've added n open and n closed parentheses, we have a valid combination
        if open_count == n and closed_count == n:
            # Build the string by popping the elements in the stack in order
            self.result.append("".join(self.stack))  # Use the stack to form the string
            return

        # Add an open parenthesis if we haven't yet added n open ones
        if open_count < n:
            self.stack.append("(")
            self.backtrack(n, open_count + 1, closed_count)
            self.stack.pop()  # Backtrack by removing the last element added to the stack

        # Add a closing parenthesis if it can balance with the number of open ones
        if closed_count < open_count:
            self.stack.append(")")
            self.backtrack(n, open_count, closed_count + 1)
            self.stack.pop()  # Backtrack by removing the last element added to the stack

